placement/cluster_mapping_provider.go

```diff
@@ -0,0 +1,476 @@
+type databaseMappings struct {
+	name               string
+	version            int32
+	maxRetentionInSecs int32
+	active             []*activeClusterMapping
+	log                xlog.Logger
+}
+
+// newDatabaseMappings returns a newly initialized set of databaseMappings
+func newDatabaseMappings(db *schema.Database, log xlog.Logger) (*databaseMappings, error) {
+	dbm := &databaseMappings{
+		name:               db.Properties.Name,
+		version:            db.Version,
+		maxRetentionInSecs: db.Properties.MaxRetentionInSecs,
+		log:                log,
+	}
+
+	for _, rules := range db.MappingRules {
+		if err := dbm.applyRules(rules); err != nil {
+			return nil, err
+		}
+	}
+
+	return dbm, nil
+}
+
+// clone clones the database mappings
+func (dbm *databaseMappings) clone() *databaseMappings {
+	clone := &databaseMappings{
+		name:               dbm.name,
+		version:            dbm.version,
+		maxRetentionInSecs: dbm.maxRetentionInSecs,
+		log:                dbm.log,
+		active:             make([]*activeClusterMapping, len(dbm.active)),
+	}
+
+	for n, m := range dbm.active {
+		var prior *clusterMapping
+		if m.prior != nil {
+			prior = m.prior.clone()
+		}
+
+		clone.active[n] = &activeClusterMapping{
+			clusterMapping: clusterMapping{
+				cluster:          m.cluster,
+				readCutoverTime:  m.readCutoverTime,
+				writeCutoverTime: m.writeCutoverTime,
+				prior:            prior,
+			},
+			shards: m.shards.Clone(),
+		}
+	}
+
+	return clone
+}
+
+// update updates the database with new mapping information
+func (dbm *databaseMappings) update(db *schema.Database) error {
+	// Short circuit if we are already at the proper version
+	if dbm.version == db.Version {
+		return nil
+	}
+
+	// Apply new rules
+	// NB(mmihic): Assumes mapping rule sets are sorted in ascending version order
+	for _, rules := range db.MappingRules {
+		if rules.ForVersion <= dbm.version {
+			continue
+		}
+
+		if err := dbm.applyRules(rules); err != nil {
+			return err
+		}
+	}
+
+	dbm.version = db.Version
+	return nil
+}
+
+// applyRules generates new mappings based on the given set of rules
+func (dbm *databaseMappings) applyRules(rules *schema.ClusterMappingRuleSet) error {
+	dbm.log.Infof("applying rule set %d for database %s", rules.ForVersion, dbm.name)
+
+	dbm.log.Infof("applying %d transitions for database %s", len(rules.ShardTransitions), dbm.name)
+	for _, t := range rules.ShardTransitions {
+		ruleShards := bitset.From(t.Shards.Bits)
+
+		// if there is no FromCluster, this is an initial assignment
+		if t.FromCluster == "" {
+			dbm.log.Infof("assigning %d shards to %s", ruleShards.Count(), t.ToCluster)
+			dbm.active = append(dbm.active, &activeClusterMapping{
+				shards: ruleShards,
+				clusterMapping: clusterMapping{
+					cluster:          t.ToCluster,
+					readCutoverTime:  xtime.FromUnixMillis(t.ReadCutoverTime),
+					writeCutoverTime: xtime.FromUnixMillis(t.WriteCutoverTime),
+				},
+			})
```

Comment:
My general rule is to only create a func / constructor when you do a bit of code more than once.  Otherwise it's just cognitive burden.